rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala

```diff
@@ -194,38 +194,37 @@ trait StreamTInstances2 {
 
   implicit def streamTSync[F[_]](
       implicit F0: Sync[F],
+      M0: Monad[StreamT[F, ?]],
       AL0: Alternative[StreamT[F, ?]]
   ): Sync[StreamT[F, ?]] =
     new StreamTSync[F]() {
       implicit val F  = F0
+      implicit val M  = M0
       implicit val AL = AL0
     }
 }
 
 private trait StreamTSync[F[_]] extends Sync[StreamT[F, ?]] with StreamTMonadError[F, Throwable] {
   implicit def F: Sync[F]
+  implicit def M: Monad[StreamT[F, ?]]
   implicit def AL: Alternative[StreamT[F, ?]]
 
   import cats.effect.ExitCase
 
   def bracketCase[A, B](acquire: StreamT[F, A])(use: A => StreamT[F, B])(
       release: (A, ExitCase[Throwable]) => StreamT[F, Unit]
   ): StreamT[F, B] =
-    StreamT.fromStream(
-      F.bracketCase[Stream[A], Stream[B]](StreamT.run(acquire)) {
-        case Stream.Empty => F.pure[Stream[B]](Stream.empty)
-        case cons: Cons[A] =>
-          StreamT
-            .run(cons.map(use).fold(StreamT.empty[F, B])(AL.combineK))
+    StreamT(
+      F.bracketCase[Step[F, A], Step[F, B]](acquire.next) {
+        case SNil() => F.pure(SNil())
+        case SCons(head, tail) => {
+          AL.combineK(use(head), M.flatMap(tail)(use)).next
```

Comment:
With obtaining the lock once, then doing the whole stream, and then releasing it, we exclude exchanging the lock usage between the elements of two streams using it simultaneously. I think it's not a problem in our case though, and the laws seem to agree.

Reply:
We really need to Nil-out the result if release returns a Nil. See the equivalent code for OptionT or Either:

```
{
            case (None, _) => F.unit //Nothing to release
            case (Some(a), ExitCase.Completed) =>
              release(a, ExitCase.Completed).value.flatMap {
                case None => ref.set(true)
                case Some(_) => F.unit
              }
            case (Some(a), res) => release(a, res).value.void
          }.flatMap[Option[B]] {
            case s @ Some(_) => ref.get.map(b => if (b) None else s)
            case None => F.pure(None)
          }
```

Reply:
Please add a test for that, and maybe see how do they test that in cats-effect

Reply:
Why do you think "we really need to Nil-out the result if release returns a Nil"?

Reply:
Please do not respond outside of the thread, it makes the conversation harder to follow.
This should by no means be a blocker for this pull request, but ok. It is implemented